### ISSUE TITLE
Autogenerated sidebar items: allow more control on output

### ISSUE THREAD

**Issue by Josh-Cena** [facebook/docusaurus#5689]
## 🚀 Feature

Several things:

- Add `exclude` option in category metadata to filter out docs;
- Add `additionalItems` option to include more items in autogenerated categories, including `link`, `ref`, etc.;
- Exclude entire subdirectories when declaring autogenerated sidebar items.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues)?

Yes

### Has this been requested on [Canny](https://docusaurus.io/feature-requests)?

No, but there are inline comments asking if they should be allowed, and the answer is yes: I've been asked about this feature.

## Motivation

Sometimes we want a fully autogenerated sidebar, but occasionally want to add a few external links in categories. Sometimes we have a legacy directory structure and we only want to generate the sidebar from part of that directory.

## API Design

In `_category_.json`, add the following options:

```diff
type CategoryMetadatasFile = {
  label?: string;
  position?: number;
  collapsed?: boolean;
  collapsible?: boolean;
  className?: string;
+ additionalItems?: WithPosition<SidebarItem>[];
+ exclude?: {
+   paths: string[];
+   docIDs?: string[];
+ };
};
```

`paths` accepts folder paths (I don't know if file paths would work well; from my experience with the autogenerator code, seems it's not easy since the doc metadata only includes `sourceDir`?), while `docIDs` accepts... doc IDs. It's assumed that the members in these two arrays are otherwise included in the category; if they are never included (non-existent IDs/paths not in the autogen dir...), maybe throw an error, or maybe do nothing.

`additionalItems` accepts `doc`, `ref`, `link`, and even `category`, but not `autogenerated` (at least I think it doesn't make much sense, and opens up holes to infinite recursion). Because `doc` items already come with their own `sidebarPosition`, they will be sorted well with the rest of the items. However, all items can have an additional `sidebarPosition` attribute (hence `WithPosition<SidebarItem>`) to override this behavior.

The `autogenerated` sidebar item will also accept the `exclude` and `additionalItems` properties, because the metadata file in the autogen dir root is not read.

## Have you tried building it?

No

**Comment by Josh-Cena** [comment]
> Would you design this API the same if autogen sidebars were not implemented?

Hard to say now I do know what's going on behind the scenes:P But even before I went down that rabbit hole, I was hoping for an `include` and `exclude` glob pattern; and in order for that to be extensible to links, I changed `include` to `additionalItems`. Hence this proposal.

> it would allow to completely override the autogen for this specific dir for which you would need to be exhaustive

What if I want some additional links but keep the rest of the 10 categories + 50 doc links auto generated?

**Comment by Josh-Cena** [comment]
My mind always explodes when I work on recursion in our sidebar structure :P E.g. do we process the category's children prior to processing itself?

I'll revisit this proposal later since we recently have/will have quite a few changes with the sidebar architecture. Maybe category metadata would be read ahead of time. After a few weeks we may be in a better position to think about the API

**Comment by lorenzolewis** [comment]
Keeping an eye on this one. Just to help with my understanding:
- I have an autogenerated sidebar slice
- Inside the directory being pointed to I have an `index.md` file
- Right now that index file is generated in the sidebar (and no way to remove this via the autogenerated sidebar type)
- With this feature, I'd be able to add an `exclude` option (or something similar) to the autogenerated sidebar object to exclude that `index.md` from being rendered as a child link.

Referencing https://github.com/tauri-apps/tauri-docs/issues/347

**Comment by slorber** [comment]
@lorenzolewis that looks like a similar case to https://github.com/facebook/docusaurus/pull/5830#issuecomment-1017146250 (except your doc is named `index`)

You should be able to do something like:

```js
[
  {
    type: "category",
    label: "JS",
    link: {
      type: "doc",
      id: "index"
    },
    items: [
      {
        type: "autogenerated",
        dirName: "js",
        exclude: ["index"]
      }
    ]
  }
];
```

Eventually, we could figure a way to exclude it automatically, as it looks like a convenient default

We could also allow an include option so that you can generate the sidebar slice one level upper:

```js
[
  {
    type: "autogenerated",
    dirName: ".",
    include: ["js"]
  }
];
```

Does it make sense?

**Comment by lorenzolewis** [comment]
@slorber makes perfect sense, thanks!

**Comment by KatherineWhan** [comment]
@slorber  I like the idea of being able to add docs too.

I have just had a read through the whole issue. Will there be any way to sort the sidebar items other than using position in the `_category_.json`? Ideally sidebar_label alpha would be perfect by default! 

I have a workflow where kb articles will be scraped into my repo from up to 20 'source' repos and deposited into the kb folder.  It is impossible to get the sidebar_position right and the current default sort is filename alpha which creates 'interesting results' as we have no control over how the teams name their files.

**Comment by slorber** [comment]
> @slorber I like the idea of being able to add docs too.

Add docs where? 

> I have just had a read through the whole issue. Will there be any way to sort the sidebar items other than using position in the _category_.json? Ideally sidebar_label alpha would be perfect by default!

> I have a workflow where kb articles will be scraped into my repo from up to 20 'source' repos and deposited into the kb folder. It is impossible to get the sidebar_position right and the current default sort is filename alpha which creates 'interesting results' as we have no control over how the teams name their files.

I don't understand what `kb` is and what you mean exactly here

Please give a very concrete example by providing a file-system tree with files/folders, the config you want to provide, and what the sidebar should look like.

Note that in any case, you can control the sidebar generation logic, including the order, by providing your own generator, or "enhancing" the existing one.

The doc provides an example where the order of the sidebar items is reversed:

https://docusaurus.io/docs/sidebar#customize-the-sidebar-items-generator


```js
// Reverse the sidebar items ordering (including nested category items)
function reverseSidebarItems(items) {
  // Reverse items in categories
  const result = items.map((item) => {
    if (item.type === 'category') {
      return {...item, items: reverseSidebarItems(item.items)};
    }
    return item;
  });
  // Reverse items at current level
  result.reverse();
  return result;
}

module.exports = {
  plugins: [
    [
      '@docusaurus/plugin-content-docs',
      {
        async sidebarItemsGenerator({defaultSidebarItemsGenerator, ...args}) {
          const sidebarItems = await defaultSidebarItemsGenerator(args);

	      // Here, it's up to you to do whatever you want to re-order the sidebar items  
          return reverseSidebarItems(sidebarItems);
        },
      },
    ],
  ],
};
```


Now it's up to you to provide your own ordering logic:
- you can create exceptions and decide to reverse only a category that has a specific label
- you can invent your own frontmatter and use `args.docs` + `sidebarItems` to compute the new order based on your custom frontmatter

**Comment by dwmkerr** [comment]
Can I suggest we also look at this comment from @slorber:

https://github.com/facebook/docusaurus/issues/3464#issuecomment-820336747

**tl;dr**

```
module.exports = {
  mySidebar: [
    {
      type: "autogenerated",
      dirName: "tutorials"
      include: "**/easy-*.md"
    },
    {
      type: "autogenerated",
      dirName: "tutorials"
      include: "**/easy-*.md"
    }
  ]
};
```

This would be *amazing* for me - my book [Effective Shell](https://github.com/dwmkerr/effective-shell) uses a folder per article:

<img width="809" alt="image" src="https://user-images.githubusercontent.com/1926984/155464131-c4d432a9-7086-45a7-b9c5-b1aa2c3fafec.png">

Unless I have missed something, this nesting structure means my sidebars have to look like this:

```
anotherSidebar: [
    'index',
    {
      type: 'category',
      label: 'Transitioning to the Shell',
      link: { type: 'doc', id: 'transitioning-to-the-shell/index'},
      items: [
        'transitioning-to-the-shell/getting-started/index',
        'transitioning-to-the-shell/navigating-your-system/index',
        'transitioning-to-the-shell/managing-your-files/index',
        'transitioning-to-the-shell/clipboard-gymnastics/index',
        'transitioning-to-the-shell/getting-help/index',
        'transitioning-to-the-shell/the-renaissance-of-the-shell/index',
      ],
    },
```

If instead I could do this:

```
anotherSidebar: [
    'index',
    {
      type: 'category',
      label: 'Transitioning to the Shell',
      link: { type: 'doc', id: 'transitioning-to-the-shell/index'},
      items: [
        'transitioning-to-the-shell/getting-started/index',
        'transitioning-to-the-shell/navigating-your-system/index',
        'transitioning-to-the-shell/managing-your-files/index',
        'transitioning-to-the-shell/clipboard-gymnastics/index',
        'transitioning-to-the-shell/getting-help/index',
        'transitioning-to-the-shell/the-renaissance-of-the-shell/index',
      ],
    },
```

If instead I could do this:

```
      type: 'category',
      label: 'Transitioning to the Shell',
      link: { type: 'doc', id: 'transitioning-to-the-shell/index'},
      include: "**/index.md"
     
```

That would be amazing

**Comment by slorber** [comment]
@dwmkerr I'm not sure to understand: what's the result you currently have, and what's the result you want?

It looks like you display twice the same sidebar in your comment, maybe a typo?

**Comment by dwmkerr** [comment]
Sorry there was mistake - copy/paste error. I've fixed it now. Because each page is in a _subfolder_, I cannot autogenerate the pages without each one sitting in it's own category. What I'd like to be able to do is have a single category, such as "Transitioning to the Shell" and have a flat set of pages in this category which are populated using a globbing pattern that finds all of the pages in the child folders

**Comment by Josh-Cena** [comment]
@dwmkerr Have you tried `beta.16`? In the latest version, a folder containing only an `index.md` should already generate a single doc link instead of a category.

**Comment by alexfornuto** [comment]
I would like to see exclude added to either `_category_.json` or sidebar configs. I have directories in /docs that I want to be excluded from the main sidebar. I link to them in the header and they use their own sidebar once loaded.

**Comment by dwmkerr** [comment]
@Josh-Cena yes I'm actually on beta 20, but still have the issues of the 'index' file being included when we auto-generate items and then also the subfolder issues, so the wildcard option plus the 'exclude' options could still be a massive improvement for me!

**Comment by Josh-Cena** [comment]
@dwmkerr Are you running into this bug? https://github.com/facebook/docusaurus/issues/7348 Try with canary?

**Comment by dwmkerr** [comment]
No I'm afraid not - I have a category but it there are also items

**Comment by Josh-Cena** [comment]
Ah, so it's the problem we are talking above. No problem—it will be the top use-case for this feature.

**Comment by azinit** [comment]
Still actual!

**Comment by lukaszch** [comment]
Is the exclude from autogenerated sidebar option already implemented? I tried it, but it doesn't work in the 2.1 version.
I have only a docs-based site, where one of the docs pages is the initial page, which works great.
I only cannot find a way to hide it from the autogenerated sidebar.

**Comment by slorber** [comment]
If this issue is not closed, this means it's not implemented not released

> I have only a docs-based site, where one of the docs pages is the initial page, which works great.
I only cannot find a way to hide it from the autogenerated sidebar.

Interesting use-case to consider, thanks for reporting

**Comment by blessanm86** [comment]
We have docs in multiple repos. So we use git submodules
Maybe I'm doing things wrong

This is our folder structure
```
docs/
├─ guides/
│  ├─ getting-started.md
│  ├─ contributing.md
├─ submodules/
│  ├─ request/
│  │  ├─ README.md
│  ├─ product-analytics/
│  │  ├─ README.md
```

I am trying to generate a sidebar structure of

```
Guides
├─ Getting Started
├─ Contributing
Request
Product Analytics
```

I am just trying to remove the submodules folder from showing up.
My sidebar config looks like this

```
{ type: 'autogenerated', dirName: '.' },
{
   type: 'doc',
   id: 'submodules/request/README',
   label: 'request,
}
```

I was hoping for something like
`{ type: 'autogenerated', dirName: '.' , exclude: ['./submodules/**/*.md']}`

**Comment by hrumhurum** [comment]
In one of the projects I have a page that should not be visible in a sidebar, but the page still should be associated with it. Allowing a sidebar label to be an empty string in frontmatter of the page perfectly communicates the intent:

```
sidebar_label: ""
```

This is not strictly an exclusion, it just makes the corresponding sidebar item invisible while keeping the page <-> sidebar association active.

A use case: some pages should not be indexable or observable in navigation because they are shared by direct links only. At the same time, those "secret" pages belong to a larger resource such as Knowledge Base, meaning that the corresponding sidebar should remain visible for such pages to allow further navigation.

**Comment by MattiasBuelens** [comment]
You can implement something like this using [a custom sidebar items generator](https://docusaurus.io/docs/sidebar/autogenerated#customize-the-sidebar-items-generator):
```js
async function postProcess({ item, ...args }) {
  if (item.type === 'category') {
    // Recurse through children
    for (const childItem of item.items) {
      await postProcess({ item: childItem, ...args });
    }
    // Add additional items
    if (item.customProps?.additionalItems) {
      for (const { position, ...additionalItem } of item.customProps.additionalItems) {
        if (position !== undefined) {
          item.items.splice(position - 1, 0, additionalItem);
        } else {
          item.items.push(additionalItem);
        }
      }
    }
  }
}

export default async function sidebarItemsGenerator({ defaultSidebarItemsGenerator, item, ...args }) {
  const sidebarItems = await defaultSidebarItemsGenerator({ item, ...args });
  for (const item of sidebarItems) {
    await postProcess({ item, defaultSidebarItemsGenerator, ...args });
  }
  return sidebarItems;
}
```

See [sidebarItemsGenerator.ts](https://github.com/THEOplayer/documentation/blob/f4dd44ed1fea8f0c58fc95a51982bfd8af584227/src/plugin/sidebarItemsGenerator.ts) for the full code, and [this `_category_.json` file](https://github.com/THEOplayer/documentation/blob/f4dd44ed1fea8f0c58fc95a51982bfd8af584227/theoplayer/getting-started/02-frameworks/_category_.json#L8-L25) for an example usage. It's not perfect (the `position` handling is very hacky), but it does the job. 😁

**Comment by voetberg** [comment]
Hello! This seems like a rather old issue but I just want to verify - the `exclude` feature in autogenerated was never implemented, correct? Is it still planned? I've been chasing around interlinked issues and couldn't see if it had come to a conclusion or not.

**Comment by Blargian** [comment]
I'd love to see a way to control ordering of sidebar items directly from the sidebar declaration. For example, I have some autogenerated items for a change log that I want to show in reverse order (e.g version 25.4 first and 22.1 last). It would be cool to be able to do something like this:

```
items: [
  {
    type: "autogenerated",
    dirName: "cloud/changelogs",
    sidebarItemsGenerator: ({ defaultSidebarItemsGenerator, ...args }) => {
      const items = defaultSidebarItemsGenerator(args);
      return items.reverse();
    }
  }
]
```

I know it's possible to do from the plugin config or from the page metadata but this seems more convenient.

**Comment by slorber** [comment]
@Blargian, the problem with that approach is that it wouldn't play well with our current docs versioning system. 

When versioning your docs, we copy and evaluate your dynamic `sidebars.js` file to a static versioned sidebar JSON fil, which can still contain autogenerated items. 

It's not possible to encode your item with a custom generator as JSON, so we'd need to entirely rethink the versioning process.

Should we make the versioned sidebars entirely static? We could run your autogenerated items and custom generator logic ahead of time to produce "final" items, but this may produce a worse DX where adding an MDX file to a versioned docs wouldn't appear automatically in the sidebar anymore

Should we emit versioned `sidebars.js` dynamic files? We could do so, but if you use any relative import in it, we must adjust the paths to the new file location, and this possibly opens the door to weird problems on complex site setups.

**Comment by Blargian** [comment]
> [@Blargian](https://github.com/Blargian), the problem with that approach is that it wouldn't play well with our current docs versioning system.
> 
> When versioning your docs, we copy and evaluate your dynamic `sidebars.js` file to a static versioned sidebar JSON fil, which can still contain autogenerated items.
> 
> It's not possible to encode your item with a custom generator as JSON, so we'd need to entirely rethink the versioning process.
> 
> Should we make the versioned sidebars entirely static? We could run your autogenerated items and custom generator logic ahead of time to produce "final" items, but this may produce a worse DX where adding an MDX file to a versioned docs wouldn't appear automatically in the sidebar anymore
> 
> Should we emit versioned `sidebars.js` dynamic files? We could do so, but if you use any relative import in it, we must adjust the paths to the new file location, and this possibly opens the door to weird problems on complex site setups.

@slorber, thanks for the explanation, that makes sense. What about supporting an `order` property in `_category_.json`? Even something as simple as controlling the order on a per folder basis would be sufficient for what I wanted to achieve. For example:

```
// _category_.json
{
  "label": "My Autogenerated Sidebar Item",
  "collapsible": true,
  "collapsed": true,
  "order": "reversed"
}
``` 

For a folder with structure:

```
-- folder
    |_ 01_a.md
    |_ 02_b.md
    |_ 03_c.md
    |_ _category_.json 
```

The menu would then display:

```
- c
- b
- a
```

**Comment by slorber** [comment]
@Blargian, I don't understand why you would prefix the document with `01_` if your goal is to make it appear last. Just prefix it with `03_` instead.

We give you:
- file name number prefixes (that is used to infer a default sidebar position)
- the `sidebar_position` front matter (that can override the number prefix)
- the `parseFrontMatter` hook that you can use to eventually generalize and automate the addition of `sidebar_position` front matter


I think that's largely enough to control the position of items.


If you really want to support your weird use case where you reverses the items, you can even implement it yourself with the `parseFrontMatter` hook. Here's a pseudo implementation to inspire you:


```ts
export default {
  markdown: {
    parseFrontMatter: async (params) => {
      const result = await params.defaultParseFrontMatter(params);

      if (!frontMatter.sidebar_position) {
        frontMatter.sidebar_position = parseNumberPrefix(params.filePath);
      }

      // "reverse" the position, if defined
      if (frontMatter.sidebar_position) {
        frontMatter.sidebar_position *= -1;
      }

      return result;
    },
  },
};
```

**Comment by bmaupin** [comment]
For excluding items I used the workaround here: https://docusaurus.canny.io/feature-requests/p/hiding-parts-of-docs-in-autogenerated-sidebar

1. Add custom CSS

   ```css
   .hiddenFromSidebar {
     display: none !important;
   }
   
   /* Hide the parent as well; this fixes the gap on hidden items when using DocCardList */
   article:has(> .hiddenFromSidebar) {
     display: none !important;
   }
   ```

1. Set the class
   - Markdown page:
     ```markdown
     ---
     sidebar_class_name: hiddenFromSidebar
     ---
     ```
   - Or a _category_.json file:
     ```json
     {
       "className": "hiddenFromSidebar"
     }
     ```

I also toyed with filtering out items using an [overridden sidebar items generator](https://docusaurus.io/docs/next/sidebar/autogenerated#customize-the-sidebar-items-generator) but it didn't seem as clean and I wasn't sure what side effects it might have:

```typescript
const formatSidebarItems = (items: NormalizedSidebar) => {
  items = items.filter((item) => {
    // Hide Archive from the sidebar and category listings
    if (item.type === 'category' && item.label === 'archive') {
      return false;
    } else {
      return true;
    }
  });
```
